File: algorithms-week10/dijkstra/cpp/dijkstra.cpp
//
// Created by Sam on 9/14/2018.
//

#include <iostream>
#include <vector>
#include <queue>
#include <climits>

using std::vector;
using std::queue;
using std::pair;
using std::priority_queue;

class indexWeight {
 public:
  int index;
  int weight;
  indexWeight(int i, int w) {
    index = i;
    weight = w;
  }
};

class weight_comp {
 public:
  bool operator() (const indexWeight& lhs, const indexWeight& rhs) const {
    return (lhs.weight > rhs.weight);
  }
};

int distance(vector<vector<int> > &adj, vector<vector<int> > &cost, int s, int t) {
  size_t n = adj.size();
  vector<int> dist(n, INT_MAX);
  priority_queue<indexWeight, vector<indexWeight>, weight_comp> pq;

  dist[s] = 0;
  pq.emplace(indexWeight(s, dist[s]));

  while (!pq.empty()) {
    indexWeight iw = pq.top();
    pq.pop();
    int v = iw.index;

    for (int i = 0; i < adj[v].size(); i++) {
      int w = adj[v][i];
      int d = dist[v] + cost[v][i];

      if (dist[w] > d) {
        dist[w] = d;
        pq.push(indexWeight(w, dist[w]));
      }
    }
  }

  if (dist[t] == INT_MAX) {
    return -1;
  }
  return dist[t];
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  vector<vector<int> > cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }
  int s, t;
  std::cin >> s >> t;
  s--, t--;
  std::cout << distance(adj, cost, s, t);
}

File: algorithms-week10/dijkstra/python/dijkstra.py
#Uses python3

import sys
import queue


def distance(adj, cost, s, t):
    n = len(adj)
    dist = [sys.maxsize] * n
    pq = queue.PriorityQueue(n)

    dist[s] = 0
    pq.put([dist[s], s])

    while not pq.empty():
        _, v = pq.get()

        for i in range(len(adj[v])):
            w = adj[v][i]
            d = dist[v] + cost[v][i]
            if dist[w] > d:
                dist[w] = d
                pq.put([dist[w], w])

    if dist[t] == sys.maxsize:
        return -1
    return dist[t]


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    data = data[3 * m:]
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for ((a, b), w) in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    s, t = data[0] - 1, data[1] - 1
    print(distance(adj, cost, s, t))
